### PR TITLE
reflect: `TypePath` part 2

### PR DESCRIPTION
**Pull request by soqb**
# Objective

- Followup to #7184.
- ~Deprecate `TypeUuid` and remove its internal references.~ No longer part of this PR.
- Use `TypePath` for the type registry, and (de)serialisation instead of `std::any::type_name`.
- Allow accessing type path information behind proxies.

## Solution
- Introduce methods on `TypeInfo` and friends for dynamically querying type path. These methods supersede the old `type_name` methods.
- Remove `Reflect::type_name` in favor of `DynamicTypePath::type_path` and `TypeInfo::type_path_table`.
- Switch all uses of `std::any::type_name` in reflection, non-debugging contexts to use `TypePath`. 

---

## Changelog

- Added `TypePathTable` for dynamically accessing methods on `TypePath` through `TypeInfo` and the type registry.
- Removed `type_name` from all `TypeInfo`-like structs.
- Added `type_path` and `type_path_table` methods to all `TypeInfo`-like structs.
- Removed `Reflect::type_name` in favor of `DynamicTypePath::reflect_type_path` and `TypeInfo::type_path`.
- Changed the signature of all `DynamicTypePath` methods to return strings with a static lifetime.

## Migration Guide

- Rely on `TypePath` instead of `std::any::type_name` for all stability guarantees and for use in all reflection contexts, this is used through with one of the following APIs:
  - `TypePath::type_path` if you have a concrete type and not a value.
  - `DynamicTypePath::reflect_type_path` if you have an `dyn Reflect` value without a concrete type.
  - `TypeInfo::type_path` for use through the registry or if you want to work with the represented type of a `DynamicFoo`.
  
- Remove `type_name` from manual `Reflect` implementations.
- Use `type_path` and `type_path_table` in place of `type_name` on `TypeInfo`-like structs.
- Use `get_with_type_path(_mut)` over `get_with_type_name(_mut)`.

## Note to reviewers

I think if anything we were a little overzealous in merging #7184 and we should take that extra care here.

In my mind, this is the "point of no return" for `TypePath` and while I think we all agree on the design, we should carefully consider if the finer details and current implementations are actually how we want them moving forward.

For example [this incorrect `TypePath` implementation for `String`](https://github.com/soqb/bevy/blob/3fea3c6c0b5719dfbd3d4230f5282ec80d82556a/crates/bevy_reflect/src/impls/std.rs#L90) (note that `String` is in the default Rust prelude) snuck in completely under the radar.

### CODE REVIEW

**Comment by soqb** [comment]
i've stripped down this PR (`TypeUuid`) is no longer deprecated, in the hopes we can squeeze this in before 0.11!

File: crates/bevy_reflect/src/type_path.rs [hunk 1]
```diff
@@ -161,3 +167,58 @@ impl<T: TypePath> DynamicTypePath for T {
         Self::module_path()
     }
 }
+
+/// Provides dynamic access to all methods on [`TypePath`].
+#[derive(Clone, Copy)]
+pub struct TypePathVtable {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/type_path.rs:173]: Nit: I wonder if there is a friendlier name than `Vtable`. Maybe `Fns` to mimic `ReflectComponentFns`? Or `Lookup`?

`Vtable` is fine haha, but could be a little intimidating.

File: crates/bevy_reflect/src/type_path.rs [hunk 2]
```diff
@@ -161,3 +167,58 @@ impl<T: TypePath> DynamicTypePath for T {
...
+    // Cache the type path as it is likely the only one that will be used.
+    type_path: &'static str,
+    short_type_path: fn() -> &'static str,
+    type_ident: fn() -> Option<&'static str>,
+    crate_name: fn() -> Option<&'static str>,
+    module_path: fn() -> Option<&'static str>,
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/type_path.rs:179]: I wonder if the increased memory footprint would worth it to cache all of these values, so as to avoid any performance cost caused by reading from a `RwLock` (on my machine doing so is a 8–11ns operation).

Probably not, right?

File: crates/bevy_reflect/src/type_path.rs [hunk 3]
```diff
@@ -161,3 +167,58 @@ impl<T: TypePath> DynamicTypePath for T {
...
+}
+
+impl fmt::Debug for TypePathVtable {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(self.type_path, f)
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/type_path.rs:184]: Nit: I feel like we should still indicate that this is a `TypePathVtable`.

File: crates/bevy_reflect/src/enums/enum_trait.rs [hunk 4]
```diff
@@ -126,15 +126,14 @@ pub trait Enum: Reflect {
     }
     /// Returns the full path to the current variant.
     fn variant_path(&self) -> String {
-        format!("{}::{}", self.type_name(), self.variant_name())
+        format!("{}::{}", self.reflect_type_path(), self.variant_name())
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/enums/enum_trait.rs:129]: Lol just occurred to me that this wouldn't return a valid path. Not really an issue (and certainly not one this PR should bother fixing). Just wanted to point out my own mistake haha.

File: crates/bevy_reflect/src/impls/std.rs [hunk 5]
```diff
@@ -87,7 +87,8 @@ impl_reflect_value!(isize(
 ));
 impl_reflect_value!(f32(Debug, PartialEq, Serialize, Deserialize, Default));
 impl_reflect_value!(f64(Debug, PartialEq, Serialize, Deserialize, Default));
-impl_reflect_value!(String(
+impl_type_path!(str);
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:90]: Interesting, why is this needed?

File: crates/bevy_reflect/src/impls/std.rs [hunk 6]
```diff
@@ -657,20 +653,12 @@ impl_reflect_for_hashmap!(::std::collections::HashMap<K, V, S>);
...
 
 impl_reflect_for_hashmap!(bevy_utils::hashbrown::HashMap<K, V, S>);
 impl_type_path!(::bevy_utils::hashbrown::hash_map::DefaultHashBuilder);
 impl_type_path!(
     ::bevy_utils::hashbrown::HashMap<K, V, S>
-    where
-        K: FromReflect + Eq + Hash + ?Sized,
-        V: FromReflect + ?Sized,
-        S: BuildHasher + Send + Sync + 'static,
 );
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:662]: Does this need to be formatted?

File: crates/bevy_reflect/src/impls/std.rs [hunk 7]
```diff
@@ -1077,21 +1055,26 @@ impl<T: FromReflect + TypePath> Typed for Option<T> {
...
     }
 }
 
-impl_type_path!(::core::option::Option<T: FromReflect + TypePath>);
+impl_type_path!(::core::option::Option<T>);
 
-impl Reflect for Cow<'static, str> {
-    fn type_name(&self) -> &str {
-        std::any::type_name::<Self>()
+impl<T: TypePath + ?Sized> TypePath for &'static T {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:1065]: Gotcha, I'm assuming the impl for `str` is because of this.

File: crates/bevy_reflect/src/lib.rs [hunk 8]
```diff
@@ -1093,28 +1093,28 @@ mod tests {
     }
 
     #[test]
-    fn dynamic_names() {
+    fn not_dynamic_names() {
         let list = Vec::<usize>::new();
         let dyn_list = list.clone_dynamic();
-        assert_eq!(dyn_list.type_name(), std::any::type_name::<Vec<usize>>());
+        assert_ne!(dyn_list.reflect_type_path(), Vec::<usize>::type_path());
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:1099]: The PR description should make a note of this in the migration guide. Users who were relying on `type_name` being that of the represented type, cannot simply change to using `reflect_type_path`.

Thankfully, it looks like you updated `Reflect::represents` so that should still work consistently.

File: crates/bevy_reflect/src/lib.rs [hunk 9]
```diff
@@ -1868,7 +1850,7 @@ bevy_reflect::tests::should_reflect_debug::Test {
             let output = to_string_pretty(&ser, config).unwrap();
             let expected = r#"
 {
-    "glam::f32::vec3::Vec3": (
+    "glam::Vec3": (
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:1853]: We did it! 🎉

File: crates/bevy_scene/src/serde.rs [hunk 10]
```diff
@@ -812,9 +812,15 @@ mod tests {
                 let received = received
                     .components
                     .iter()
-                    .find(|component| component.type_name() == expected.type_name())
+                    .find(|component| {
+                        component.get_represented_type_info().unwrap().type_path()
+                            == expected.get_represented_type_info().unwrap().type_path()
```
Comment (MrGVSV) [inline crates/bevy_scene/src/serde.rs:817]: Minor optimization: Can we move `expected.get_represented_type_info()` out of this closure to prevent calling it repeatedly?

**Review by MrGVSV — COMMENTED** [review]
Looking great! And yeah removing the `TypeUuid` changes definitely helped make this review a bit nicer.

Just some comments, questions, and nitpicks!

File: crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs [hunk 11]
```diff
@@ -35,7 +35,7 @@ pub(crate) enum ReflectDerive<'a> {
 /// //                          traits
 /// //        |----------------------------------------|
 /// #[reflect(PartialEq, Serialize, Deserialize, Default)]
-/// //            type_name       generics
+/// //            type_path       generics
```
[the remark this answers is not in the document]
Comment (soqb) [reply crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:38]: i suppose it could be seen as either, but isn't the graph supposed to show what the fields of `ReflectMeta` represent, and there isn't a `type_name` field.

File: crates/bevy_reflect/src/array.rs [hunk 12]
```diff
@@ -133,11 +143,11 @@ impl ArrayInfo {
...
 
-    /// The [type name] of the array item.
+    /// A representation of the type path of the array item.
     ///
-    /// [type name]: std::any::type_name
-    pub fn item_type_name(&self) -> &'static str {
-        self.item_type_name
+    /// Provides dynamic access to all methods on [`TypePath`].
+    pub fn item_type_path_vtable(&self) -> &TypePathVtable {
+        &self.item_type_path
```
[the remark this answers is not in the document]
Comment (soqb) [reply crates/bevy_reflect/src/array.rs:150]: its niche and trivial enough that it can be left to followup work, imo

File: crates/bevy_reflect/src/impls/std.rs (hunk 5 above)
Comment (soqb) [reply crates/bevy_reflect/src/impls/std.rs:90]: for `&'static str` and `Cow<'static, str>` since they have blanket implementations of `TypePath` with `?Sized` bounds.

File: crates/bevy_reflect/src/reflect.rs [hunk 13]
```diff
@@ -73,9 +73,6 @@ pub enum ReflectOwned {
 /// [derive macro]: bevy_reflect_derive::Reflect
 /// [crate-level documentation]: crate
 pub trait Reflect: DynamicTypePath + Any + Send + Sync {
-    /// Returns the [type name][std::any::type_name] of the underlying type.
-    fn type_name(&self) -> &str;
```
[the remark this answers is not in the document]
Comment (soqb) [reply crates/bevy_reflect/src/reflect.rs:77]: iiuc, its almost functionally equivalent to a rename `Reflect::type_name` to `DynamicTypePath::reflect_type_path`; `DynamicTypePath` doesn't even need to be in scope to call `my_reflect.reflect_type_path()`, but i suppose we could give it a deprecated default impl that calls `reflect_type_path`.

File: crates/bevy_reflect/src/type_path.rs (hunk 2 above)
Comment (soqb) [reply crates/bevy_reflect/src/type_path.rs:179]: these are pretty much never called and are just there for completeness so i don't think it would be worth it.

`short_type_path` is the only one that i think is ever called internally.

File: crates/bevy_scene/src/serde.rs (hunk 10 above)
Comment (soqb) [reply crates/bevy_scene/src/serde.rs:817]: its a test-only fn so its not performance critical but i don't see why not!

**Comment by soqb** [comment]
failure in CI fixed by #8957.

File: crates/bevy_scene/src/serde.rs (hunk 10 above)
Comment (MrGVSV) [reply crates/bevy_scene/src/serde.rs:817]: Lol my bad I completely missed the context of this being a test

File: crates/bevy_reflect/src/reflect.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MrGVSV) [reply crates/bevy_reflect/src/reflect.rs:77]: Yeah a deprecation would be good if we can make it work (I bring it up because we should probably start using it more where we can). If not, or if it's too much trouble, then we can just skip it

File: crates/bevy_reflect/src/type_path.rs (hunk 1 above)
Comment (cart) [reply crates/bevy_reflect/src/type_path.rs:173]: Yeah "vtable" is a language implementation detail imo, not a name for a high level type.

Comment (cart) [reply crates/bevy_reflect/src/type_path.rs:173]: This isn't about the Fns imo. This is about the data it provides (static type path impl strings). The fact that we're holding fns at all seems like an implementation detail to me. This is "an owned type that provides TypePath data for TypeInfo". It is effectively a DynamicTypePath impl in its own right (I'm assuming we aren't actually impling that to preserve the static lifetimes).

Things like TypePathProvider, TypePathData, TypeInfoTypePath, etc seem like the right direction.

File: crates/bevy_scene/src/scene.rs [hunk 14]
```diff
@@ -77,11 +77,11 @@ impl Scene {
                 type_registry
                     .get(type_id)
                     .ok_or_else(|| SceneSpawnError::UnregisteredType {
-                        type_name: component_info.name().to_string(),
+                        type_path: component_info.name().to_string(),
```
Comment (cart) [inline crates/bevy_scene/src/scene.rs:80]: component_info.name() does not return the `type_path`

**Comment by cart** [comment]
Still some open questions and merge conflicts here, so I'm moving this to 0.12.  This becomes most relevant for the asset system and scene system, which have more focus in 0.12 anyway.

File: crates/bevy_reflect/src/lib.rs [hunk 15]
```diff
@@ -1213,62 +1222,57 @@ mod tests {
...
-        assert_eq!(MacroName::short_type_path(), "MyMacroName");
-        assert_eq!(MacroG::<Param, 10>::short_type_path(), "MacroG<Param, 10>");
-        assert_eq!(
-            MacroNameG::<Param>::short_type_path(),
-            "MyMacroNameG<Param>"
-        );
+        assert_type_paths! {
+            u8 => "u8", "u8",
+            Type => "Long", "Short",
+            &Type => "&Long", "&Short",
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:1258]: Just to clarify, this wouldn't work for `&&Type` since `&'static Type` doesn't implement `Reflect`, right?

File: crates/bevy_reflect/src/impls/std.rs [hunk 16]
```diff
@@ -1028,21 +1002,26 @@ impl<T: FromReflect + TypePath> Typed for Option<T> {
...
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:1012]: For completeness, should we also add an impl for `&'static mut`?

File: crates/bevy_scene/src/dynamic_scene.rs [hunk 17]
```diff
@@ -73,14 +73,19 @@ impl DynamicScene {
...
+            let type_info = resource.get_represented_type_info().ok_or_else(|| {
+                SceneSpawnError::NoRepresentedType {
+                    type_path: resource.reflect_type_path().to_string(),
+                }
+            })?;
+            let registration = type_registry.get(type_info.type_id()).ok_or_else(|| {
+                SceneSpawnError::UnregisteredButReflectedType {
+                    type_path: type_info.type_path().to_string(),
+                }
+            })?;
```
Comment (MrGVSV) [inline crates/bevy_scene/src/dynamic_scene.rs:85]: Nit: Should these be let-else statements? I'm not sure which is more idiomatic haha

File: crates/bevy_scene/src/scene_spawner.rs [hunk 18]
```diff
@@ -43,12 +43,23 @@ pub struct SceneSpawner {
...
-    UnregisteredType { type_name: String },
+    #[error("scene contains the unregistered component `{type_path}`. consider adding `#[reflect(Component)]` to your type")]
+    UnregisteredComponent { type_path: String },
+    #[error("scene contains the unregistered resource `{type_path}`. consider adding `#[reflect(Resource)]` to your type")]
+    UnregisteredResource { type_path: String },
+    #[error(
+        "scene contains the unregistered type `{std_type_name}`. \
+        consider reflecting it with `#[derive(Reflect)]` \
+        and registering the type using `app.register_type::<T>()`"
+    )]
```
Comment (MrGVSV) [inline crates/bevy_scene/src/scene_spawner.rs:54]: Interesting. In what situations would this occur?

**Review by MrGVSV — APPROVED** [review]
Looks good to me!

File: crates/bevy_scene/src/scene_spawner.rs (hunk 18 above)
Comment (soqb) [reply crates/bevy_scene/src/scene_spawner.rs:54]: this error is for when we have a component id which doesn't correspond to a type in the registry. we know it hasn't been `register_type`'d but it also likely it's not even been `Reflect`ed, so i put both suggestions down.

Comment (MrGVSV) [reply crates/bevy_scene/src/scene_spawner.rs:54]: How would it be in the registry if it's not `Reflect`?

File: crates/bevy_reflect/src/serde/ser.rs [hunk 19]
```diff
@@ -667,7 +665,7 @@ mod tests {
 
         let output = ron::ser::to_string_pretty(&serializer, config).unwrap();
         let expected = r#"{
-    "bevy_reflect::serde::ser::tests::should_serialize_option::OptionTest": (
+    "bevy_reflect::serde::ser::tests::OptionTest": (
```
Comment (kristoff3r) [inline crates/bevy_reflect/src/serde/ser.rs:668]: Shouldn't the function name be part of the type path when the type is declared inside it?

**Review by kristoff3r — APPROVED** [review]
I read through everything and it looks very reasonable to me (although I'm completely new to `bevy_reflect` so take that with a grain of salt). All the changed code looks the same or better.

Also the link to the "incorrect type path implementation" is dead, it would've been nice to know what that was so I could have looked for similar stuff or missing tests.

There were some names in the changelog that hasn't been updated to match the code:
* `TypePathVtable` -> `TypePathTable`
* `type_path_vtable` -> `type_path_table`

File: crates/bevy_reflect/src/serde/ser.rs (hunk 19 above)
Comment (MrGVSV) [reply crates/bevy_reflect/src/serde/ser.rs:668]: Great question! Unfortunately, this is a limitation with `TypePath` since it relies on the Rust macro `module_path` to automatically generate the type path (and that particular macro does not take function names into account)

**Comment by alice-i-cecile** [comment]
@MrGVSV, can you go through and resolve conversations that are no longer relevant? @soqb, can you rebase? I think we can get this done for 0.12, so let's get the final steps rolling.

**Comment by MrGVSV** [comment]
> @MrGVSV, can you go through and resolve conversations that are no longer relevant?

I think this would have to be done by the author or the PR (@soqb).

**Comment by soqb** [comment]
i'll get round to updating everything this week when i get a spot

File: crates/bevy_reflect/src/lib.rs [hunk 20]
```diff
@@ -1160,18 +1160,27 @@ mod tests {
...
-            dyn_tuple_struct.type_name(),
-            std::any::type_name::<TestTupleStruct>()
+        assert_ne!(
+            dyn_tuple_struct.reflect_type_path(),
+            TestTupleStruct::type_path()
         );
     }
 
+    macro_rules! assert_type_paths {
+        ($($ty:ty => $long:literal, $short:literal,)*) => {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:1176]: Total nitpick that definitely is not required: I think wrapping the literals in parentheses (like a tuple) might be easier to read, but that's definitely just my opinion haha

File: crates/bevy_reflect/src/lib.rs [hunk 21]
```diff
@@ -1807,10 +1793,10 @@ bevy_reflect::tests::should_reflect_debug::Test {
     map: {
         123: 1.23,
     },
-    a_struct: bevy_reflect::tests::should_reflect_debug::SomeStruct {
+    a_struct: bevy_reflect::tests::SomeStruct {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/lib.rs:1796]: Hm, does `TypePath` mention this quirk (due to `module_path!`)? If not, that might be worth adding as part of this PR so users who are defining structs inside functions aren't wholly caught off guard.

File: crates/bevy_reflect/src/reflect.rs [hunk 22]
```diff
@@ -279,8 +307,10 @@ impl dyn Reflect {
     ///
     /// Read `is` for more information on underlying values and represented types.
     #[inline]
-    pub fn represents<T: Reflect>(&self) -> bool {
-        self.type_name() == any::type_name::<T>()
+    pub fn represents<T: Reflect + TypePath>(&self) -> bool {
+        self.get_represented_type_info()
+            .map(|t| t.type_path() == T::type_path())
+            .unwrap_or(false)
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/reflect.rs:313]: Should this be using `DynamicTypePath`? Or does that not work for `dyn Reflect`?
```suggestion
        self.reflect_type_path() == T::type_path()
```

Comment (soqb) [reply crates/bevy_reflect/src/reflect.rs:313]: that would give it the same behaviour as `is`. it would not respect the represented types of `DynamicFoo`s.

Comment (MrGVSV) [reply crates/bevy_reflect/src/reflect.rs:313]: Oh right. Good catch!

**Comment by MrGVSV** [comment]
> @soqb @MrGVSV It looks like there's a bit of cleanup left (and merge conflicts). Do you feel like we should block on that, or just merge this and move on?

Not sure about the merge conflicts, but I believe my remaining comments are more nits than anything we should block on.
